### PR TITLE
Move expected tracer.events from test code to golden files

### PR DESCRIPTION
**Pull request by victorjzsun**
Addresses #125. Replaces certain assertions to instead be checked using golden files.

### CODE REVIEW

**Comment by laike9m** [comment]
Generating trace.events JSON from the existing json data files is fine, but eventually we need to be able to dump these JSON from the trace.events in memory. Imagine we add a field to the events, and we definitely don't want to manually update the golden files. We already [support](https://github.com/laike9m/Cyberbrain/blob/master/test/conftest.py#L57-L66) the golden file override feature for mock response, and it would be the same for trace.events.

**Comment by laike9m** [comment]
I'm working on #126 to support Python 3.10. 3.10 changes bytecode quite a bit, and hard-code `trace.events` in test doesn't really work. I'm looking forward to this feature😀

Btw, do you think `get_value` in `test/utils.py` can be removed once this is in?

**Comment by victorjzsun** [comment]
> I'm working on #126 to support Python 3.10. 3.10 changes bytecode quite a bit, so hard-code `trace.events` in test doesn't work anymore. I'm looking forward to this feature😀
> 
> Btw, do you think `get_value` in `test/utils.py` can be removed once this is in?

`get_value` is also used in `tracer.loops` assertions so not yet. I was planning on just focusing on `tracer.events` for this PR and deal with `tracer.loops` assertions in the next PR.

File: cyberbrain/basis.py [hunk 1]
```diff
@@ -64,6 +64,8 @@ class Event:
 
     @staticmethod
     def value_serializer(inst: type, field: attr.Attribute, value: Any):
+        if field is None:
```
Comment (laike9m) [inline cyberbrain/basis.py:67]: why do we need this?

File: test/conftest.py [hunk 2]
```diff
@@ -73,4 +126,16 @@ def mocked_responses(request):
         if get_os_type() == "windows" and frame_name in {"test_numpy", "test_pandas"}:
             return
 
-        assert frame_data == golden_frame_data, json.dumps(frame_data, indent=4)
+        if golden_frame_data.get("response", None) is None:
```
Comment (laike9m) [inline test/conftest.py:129]: > If default is not given, it defaults to None
so `None` is not needed here

File: test/conftest.py [hunk 3]
```diff
@@ -40,16 +41,68 @@ def fixture_trace(request):
...
+            json.dump(golden_frame_data, f, ensure_ascii=False, indent=4)
+
+    assert tracer_events == golden_frame_data["tracer.events"], json.dumps(
+        tracer_events, indent=4
+    )
+
+
 @pytest.fixture
-def mocked_responses(request):
+def check_response(request):
```
Comment (laike9m) [inline test/conftest.py:97]: I feel like this function is doing multiple things and we should break into smaller ones, something like

```python
golden_response = get_golden_response()
if golden_response:
  assert ...
else:
  update_golden_response()
```

This makese the intent more clear. A good side effect: skip assertion if golden_response does not exist (now assertion still happens if the file exists).

File: test/conftest.py [hunk 4]
```diff
@@ -40,16 +41,68 @@ def fixture_trace(request):
...
 
+@pytest.fixture(scope="function")
+def check_tracer_events():
+    def serialize_symbol(symbol):
+        if symbol.snapshot is None:
+            return {"name": symbol.name, "snapshot": symbol.snapshot}
+        snapshot = {
+            "location": symbol.snapshot.location,
+            "events_pointer": symbol.snapshot.events_pointer,
+        }
```
Comment (laike9m) [inline test/conftest.py:52]: you can do

```python
# If snap shot is None, keep it.
snapshot = snapshot and {
    "location": symbol.snapshot.location,
    "events_pointer": symbol.snapshot.events_pointer,
}
```

File: test/conftest.py [hunk 5]
```diff
@@ -40,16 +41,68 @@ def fixture_trace(request):
...
+        return {"name": symbol.name, "snapshot": snapshot}
+
+    yield
+
+    tracer_events = []
```
Comment (laike9m) [inline test/conftest.py:57]: this can be extracted to a separate function.

File: test/conftest.py (hunk 3 above)
Comment (laike9m) [reply test/conftest.py:97]: Same with `check_tracer_events`

**Review by laike9m — CHANGES_REQUESTED** [review]
left some comments

**Comment by victorjzsun** [comment]
> Nvm, I just find that I can squash myself

Only collaborators have merge access so I can't even merge it regardless.

**Pull request merged**